### PR TITLE
[SPARK-34012][SQL] Keep behavior consistent when conf `spark.sql.legacy.parser.havingWithoutGroupByAsWhere` is true with migration guide

### PR DESCRIPTION
**Pull request by AngersZhuuuu**
### What changes were proposed in this pull request?
In https://github.com/apache/spark/pull/22696 we support HAVING without GROUP BY means global aggregate
But since we treat having as Filter before, in this way will cause a lot of analyze error, after https://github.com/apache/spark/pull/28294 we use `UnresolvedHaving` to instead `Filter` to solve such problem, but break origin logical about treat `SELECT 1 FROM range(10) HAVING true` as `SELECT 1 FROM range(10) WHERE true`   .
This PR fix this issue and add UT.

### Why are the changes needed?
Keep consistent behavior of migration guide.

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
added UT

### CODE REVIEW

**Comment by AngersZhuuuu** [comment]
FYI @maropu @cloud-fan @viirya

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala [hunk 1]
```diff
@@ -714,7 +728,7 @@ class AstBuilder extends SqlBaseBaseVisitor[AnyRef] with SQLConfHelper with Logg
     val withProject = if (aggregationClause == null && havingClause != null) {
       if (conf.getConf(SQLConf.LEGACY_HAVING_WITHOUT_GROUP_BY_AS_WHERE)) {
         // If the legacy conf is set, treat HAVING without GROUP BY as WHERE.
-        withHavingClause(havingClause, createProject())
+        withHavingClauseAsWhere(havingClause, createProject())
```
Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala:731]: Could you inline this func here?

**Comment by maropu** [comment]
Could you file a new jira for this issue?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala (hunk 1 above)
Comment (AngersZhuuuu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala]: > Could you inline this func here?

DONE

**Comment by AngersZhuuuu** [comment]
> Could you file a new jira for this issue?

DONE

**Review by cloud-fan — APPROVED** [review]
good catch! Which version did we break this legacy config?

**Comment by AngersZhuuuu** [comment]
> good catch! Which version did we break this legacy config?

SInce https://github.com/apache/spark/pull/28294 backport to 2.4 too, impacted version. should be 2.4/3.0/3.1/3.2

**Review by maropu — APPROVED** [review]

**Review by viirya — APPROVED** [review]

**Comment by maropu** [comment]
GA passed, so merged to master/branch-3.1. Could you make PRs for backporting this fix into branch-3.0/2.4?

**Pull request closed without merging**